File: decrypt.py
#!/usr/bin/env python3
"""
Patched decrypt_content_details with diagnostics.

Drop-in replacement for the function of the same name in bot.py.
Only this function is touched — nothing else in bot.py needs to change.
"""
import sys
import json
import base64
import logging
from Crypto.Cipher import AES
from Crypto.Util.Padding import unpad

logger = logging.getLogger(__name__)

# --- Config: key/IV pulled from the client bundle ---
KEY = b"Ch@tS3cr3tK3y!16"   # 16 bytes -> AES-128
IV = b"Ch@tIV#16Bytes!!"    # 16 bytes


class DecryptDiagnosticError(Exception):
    """Raised with extra context when decryption produces unusable output."""
    pass


def decrypt_content_details(base64_ciphertext, content_id=None):
    """
    Decrypts a base64-encoded AES-128-CBC ciphertext.
    Returns the parsed JSON object.

    Raises DecryptDiagnosticError with detailed context on failure,
    instead of letting json.loads raise an opaque
    "Expecting value: line 1 column 1 (char 0)" error.
    """
    label = f"content_id={content_id}" if content_id is not None else "content_id=<unknown>"

    # --- Step 1: base64 decode ---
    if not base64_ciphertext:
        raise DecryptDiagnosticError(
            f"[{label}] base64_ciphertext is empty or None before decoding."
        )

    try:
        ciphertext = base64.b64decode(base64_ciphertext)
    except Exception as e:
        raise DecryptDiagnosticError(
            f"[{label}] base64 decode failed: {e}. "
            f"Input length={len(base64_ciphertext)} chars."
        ) from e

    # --- Step 2: sanity-check ciphertext length ---
    if len(ciphertext) == 0:
        raise DecryptDiagnosticError(
            f"[{label}] Decoded ciphertext is empty (0 bytes). "
            f"The stored/base64 value for this record is likely missing or truncated."
        )

    if len(ciphertext) % AES.block_size != 0:
        raise DecryptDiagnosticError(
            f"[{label}] Ciphertext length ({len(ciphertext)} bytes) is not a "
            f"multiple of the AES block size ({AES.block_size}). "
            f"This payload is truncated/corrupted and was not encrypted with "
            f"AES-CBC using this block size."
        )

    # --- Step 3: AES-CBC decrypt ---
    try:
        cipher = AES.new(KEY, AES.MODE_CBC, IV)
        decrypted = cipher.decrypt(ciphertext)
    except Exception as e:
        raise DecryptDiagnosticError(
            f"[{label}] AES-CBC decrypt call failed: {e}"
        ) from e

    # --- Step 4: unpad, with visibility into what's being stripped ---
    try:
        plaintext_bytes = unpad(decrypted, AES.block_size)
    except Exception as e:
        raise DecryptDiagnosticError(
            f"[{label}] Padding removal failed: {e}. "
            f"Last block (hex)={decrypted[-AES.block_size:].hex()}. "
            f"This strongly suggests the KEY/IV do not match this payload."
        ) from e

    if len(plaintext_bytes) == 0:
        raise DecryptDiagnosticError(
            f"[{label}] Decryption + unpad succeeded but produced an EMPTY "
            f"plaintext (0 bytes after stripping padding). "
            f"Decrypted length before unpad={len(decrypted)} bytes. "
            f"This is the signature of a key/IV mismatch: the cipher runs "
            f"without error, but the output is garbage that happens to look "
            f"like a full block of valid padding."
        )

    # --- Step 5: decode + JSON parse ---
    try:
        plaintext = plaintext_bytes.decode("utf-8")
    except UnicodeDecodeError as e:
        raise DecryptDiagnosticError(
            f"[{label}] UTF-8 decode failed after unpad: {e}. "
            f"Plaintext byte length={len(plaintext_bytes)}. "
            f"First 32 bytes (hex)={plaintext_bytes[:32].hex()}. "
            f"Likely a key/IV mismatch producing non-UTF-8 garbage."
        ) from e

    try:
        return json.loads(plaintext)
    except json.JSONDecodeError as e:
        raise DecryptDiagnosticError(
            f"[{label}] JSON parse failed: {e}. "
            f"Plaintext length={len(plaintext)} chars. "
            f"Plaintext preview (first 80 chars)={plaintext[:80]!r}."
        ) from e


def main():
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} \"<base64_ciphertext>\"")
        sys.exit(1)

    ciphertext = sys.argv[1]
    try:
        result = decrypt_content_details(ciphertext)
        print("✅ Decryption succeeded:\n")
        print(json.dumps(result, indent=2, ensure_ascii=False))
    except DecryptDiagnosticError as err:
        print(f"❌ Decryption failed with diagnostics:\n{err}\n")
        sys.exit(1)
    except Exception as err:
        print(f"❌ Unexpected error: {err}\n")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: decrypt.js
/**
 * decrypt_test.js
 *
 * Quick utility to test-decrypt "content-details" style AES-128-CBC
 * encrypted payloads using the hardcoded key/IV found in the client bundle.
 *
 * Usage:
 *   node decrypt_test.js "<base64_ciphertext>"
 *
 * If no argument is passed, it falls back to the SAMPLE value below
 * so you can quickly sanity-check the script still works.
 */

const crypto = require('crypto');

// --- Config: key/IV pulled from the client bundle ---
const KEY = Buffer.from('Ch@tS3cr3tK3y!16'); // 16 bytes -> AES-128
const IV  = Buffer.from('Ch@tIV#16Bytes!!'); // 16 bytes

// Optional: a known-good sample so you can verify the script works
// even without passing an argument.
const SAMPLE = 'oGjN6hGaPPR9yhYWT4IN6V5YLK8jsT+R9zLxlR+YIRhu56e3ClRvfZk6KoBf3k5Rwdl/GKmeyWI1JR+qJcWH50V9xBhan1Yy8arc2w68YLc5jX0LwXiiw64dNAjSVynHHxX+SJxqqBQRfM2EybWvp1JhHvCtOM1hwGiMhIH2NxNriIM7zdBjqT0bgCWpAFnZ6DFguGWap/uHwWO+0fjhGIRhiZ3blp/8XLl63ONKeYdKSCjI2AIB/BqYD5cS/JMlZvo+JbA9F7y2Ngp6pAl8U49vu6DT62bVMvaSCsiuNq+2hxig3z3SlNYqb3FpRHaZ/SPdOOIPlwkIxPHZ2WbCMGpnfSTQVjbuFOpEHmySDiad761mYFQrJX8EM6CFiym+1JOUvodHB3cPYDLQ2jjiqJsGVO4Fewzu3K0lw7xTy2wRIUEE7zgsmiGLTttNsQXv1R7/0JCl9eoK4qvbJFJOI6U0XUeczrHET3Tv/2jNj4CLxGZAjbO5tsHGvO7EVnmRPEZOmaOG3dCo4rmS3/XA2eHRSol+JkW35d1NTeZRMVDsvURVZPKsqCOKufxC5GBPQ14RaR7rUW0n7hp6pCo83JIP/hYczszYsVPkihi64BuzHMBTOhayRU+SKIGlppnez2RjDDz4wVxUV8pEnbm98mWHRmqh+joYy8VcUETAIFD06nshO0AYVXzPsTK9HT8aALl/39RErTBH9NIbE0EZa10QvmzeX4wAznckv3lnYXpw0uhAsKNT+w+jCnsJR985gVy61yLp9Tns8t/Tl0AjRSfTQ+HmcG/VD1G5r8OhsIdD2QXxIBcYoLnpf1zKeosW2xPW30ChoWRMDOgiz5DlPhdw6lTGbrMqtmcFJ4Atg2X5qBnBxfuJIaR+0iJw9PGZWftTPJAvXe/1D4dTGNdC8b3mUjWNKtqOIj+yPtw8asQhXUN37hm0vw/uWjf0BiQS';

/**
 * Decrypts a base64-encoded AES-128-CBC ciphertext string.
 * Returns the parsed JSON object, or throws on failure.
 */
function decryptContentDetails(base64Ciphertext) {
  const decipher = crypto.createDecipheriv('aes-128-cbc', KEY, IV);
  const decrypted = Buffer.concat([
    decipher.update(Buffer.from(base64Ciphertext, 'base64')),
    decipher.final(),
  ]);
  const text = decrypted.toString('utf8');
  return JSON.parse(text);
}

// --- Run ---
const input = process.argv[2] || SAMPLE;

if (!process.argv[2]) {
  console.log('(No argument passed — using built-in SAMPLE for a sanity check)\n');
}

try {
  const result = decryptContentDetails(input);
  console.log('✅ Decryption succeeded:\n');
  console.log(JSON.stringify(result, null, 2));
} catch (err) {
  console.error('❌ Decryption failed:', err.message);
  console.error('\nPossible causes:');
  console.error('  - The string isn\'t valid base64 / was truncated when copy-pasted');
  console.error('  - The key/IV don\'t match this particular payload (e.g. different endpoint or rotated key)');
  console.error('  - The payload wasn\'t actually encrypted with AES-128-CBC');
    }
